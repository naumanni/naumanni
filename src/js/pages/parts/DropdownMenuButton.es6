import React from 'react'
import PropTypes from 'prop-types'


/**
 * メニューがにょっと出てくる
 */
export default class DropdownMenuButton extends React.Component {
  static propTypes = {
    modifier: PropTypes.string,
    // children
    onRenderMenu: PropTypes.func.isRequired,
  }

  state = {
    isEnteredMenu: false,
    isMenuVisible: false,
  }

  /**
   * @override
   */
  render() {
    const {modifier} = this.props
    let className = 'dropdownMenuButton'

    if(modifier) {
      className += ` dropdownMenuButton--${modifier}`
    }

    return (
      <div className={className} onMouseLeave={::this.onMouseLeave} style={this.props.style || {}}>
        <div className="dropdownMenuButton-button" onClick={::this.onClickButton}>
          {this.props.children}
        </div>
        {this.state.isMenuVisible && (
          <div
            className="dropdownMenuButton-menu"
            onClick={::this.onClickMenu}
            onMouseEnter={::this.onEnteredMenu}
          >
            {this.props.onRenderMenu()}
          </div>
        )}
      </div>
    )
  }

  reset() {
    this.setState({
      isEnteredMenu: false,
      isMenuVisible: false,
    })
  }

  onClickButton(e) {
    e.preventDefault()
    this.setState({isMenuVisible: !this.state.isMenuVisible})
  }

  onClickMenu() {
    this.reset()
  }

  onEnteredMenu() {
    this.setState({isEnteredMenu: true})
  }

  onMouseLeave() {
    this.state.isEnteredMenu && this.reset()
  }
}
