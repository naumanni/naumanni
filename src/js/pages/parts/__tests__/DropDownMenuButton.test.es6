import React from 'react'
import {shallow} from 'enzyme'
import serializer from 'enzyme-to-json/serializer'

import DropdownMenuButton from '../DropdownMenuButton'


expect.addSnapshotSerializer(serializer)

const toggleMenu = (wrapper) => {
  wrapper.find('.dropdownMenuButton-button').simulate('click', {
    preventDefault: jest.fn(),
  })
}

describe('DropdownMenuButton', () => {
  const button = <button className="testbutton">button</button>
  const menu = <div className="testmenu">menu</div>
  const component = (
     <DropdownMenuButton onRenderMenu={() => menu}>
       {button}
     </DropdownMenuButton>
  )
  let wrapper

  beforeEach(() => {
    wrapper = shallow(component)
  })

  afterEach(() => {
    expect(wrapper).toMatchSnapshot()
  })

  it('only button children is visible as default', () => {
    expect(wrapper.find('.testbutton').length).toBe(1)
    expect(wrapper.find('.testmenu').length).toBe(0)
  })
  it('menu is visible after button clicked', () => {
    toggleMenu(wrapper)
    expect(wrapper.find('.testmenu').length).toBe(1)
  })

  describe(':after menu shown -', () => {
    beforeEach(() => {
      toggleMenu(wrapper)
    })

    it('menu is still visible when detect mouseleave from component', () => {
      expect(wrapper.find('.testmenu').length).toBe(1)
      wrapper.find('.dropdownMenuButton').simulate('mouseleave')
      expect(wrapper.find('.testmenu').length).toBe(1)
    })
    it('menu is invisible when detect mouseleave from component after detect mouseenter from menu element', () => {
      expect(wrapper.find('.testmenu').length).toBe(1)
      wrapper.find('.dropdownMenuButton-menu').simulate('mouseenter')
      wrapper.find('.dropdownMenuButton').simulate('mouseleave')
      expect(wrapper.find('.testmenu').length).toBe(0)
    })
    it('menu is invisible after menu button clicked again', () => {
      expect(wrapper.find('.testmenu').length).toBe(1)
      toggleMenu(wrapper)
      expect(wrapper.find('.testmenu').length).toBe(0)
    })
  })
})
