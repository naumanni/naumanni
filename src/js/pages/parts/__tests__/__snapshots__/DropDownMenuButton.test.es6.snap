// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenuButton :after menu shown - menu is invisible after menu button clicked again 1`] = `
<div
  className="dropdownMenuButton"
  onMouseLeave={[Function]}
  style={Object {}}
>
  <div
    className="dropdownMenuButton-button"
    onClick={[Function]}
  >
    <button
      className="testbutton"
    >
      button
    </button>
  </div>
</div>
`;

exports[`DropdownMenuButton :after menu shown - menu is invisible when detect mouseleave from component after detect mouseenter from menu element 1`] = `
<div
  className="dropdownMenuButton"
  onMouseLeave={[Function]}
  style={Object {}}
>
  <div
    className="dropdownMenuButton-button"
    onClick={[Function]}
  >
    <button
      className="testbutton"
    >
      button
    </button>
  </div>
</div>
`;

exports[`DropdownMenuButton :after menu shown - menu is still visible when detect mouseleave from component 1`] = `
<div
  className="dropdownMenuButton"
  onMouseLeave={[Function]}
  style={Object {}}
>
  <div
    className="dropdownMenuButton-button"
    onClick={[Function]}
  >
    <button
      className="testbutton"
    >
      button
    </button>
  </div>
  <div
    className="dropdownMenuButton-menu"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="testmenu"
    >
      menu
    </div>
  </div>
</div>
`;

exports[`DropdownMenuButton menu is visible after button clicked 1`] = `
<div
  className="dropdownMenuButton"
  onMouseLeave={[Function]}
  style={Object {}}
>
  <div
    className="dropdownMenuButton-button"
    onClick={[Function]}
  >
    <button
      className="testbutton"
    >
      button
    </button>
  </div>
  <div
    className="dropdownMenuButton-menu"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="testmenu"
    >
      menu
    </div>
  </div>
</div>
`;

exports[`DropdownMenuButton only button children is visible as default 1`] = `
<div
  className="dropdownMenuButton"
  onMouseLeave={[Function]}
  style={Object {}}
>
  <div
    className="dropdownMenuButton-button"
    onClick={[Function]}
  >
    <button
      className="testbutton"
    >
      button
    </button>
  </div>
</div>
`;
